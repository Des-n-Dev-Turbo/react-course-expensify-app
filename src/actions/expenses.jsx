import { v4 as uuidv4 } from 'uuid';

//ADDEXPENSE
export const addExpense = (
    {
        description = '',
        note = '',
        amount = 0,
        createdAt = 0
    } = {}) => ({
    type: 'ADD_EXPENSE',
    expenses: {
        id: uuidv4(),
        description,
        note,
        amount,
        createdAt
    }
});

//REMOVEEXPENSE
export const removeExpense = ({ id }) => ({
    type: 'REMOVE_EXPENSE',
    id
});

//EDITEXPENSE
export const editExpense = (id, updates) => ({
    type: 'EDIT_EXPENSE',
    id,
    updates

});
