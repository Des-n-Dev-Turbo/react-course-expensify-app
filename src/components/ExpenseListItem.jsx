import React from "react";
import { Link } from "react-router-dom";
import moment from "moment";
import numeral from "numeral";

const ExpenseListItem = ({ description, id, createdAt, amount }) => (
  <Link to={`/edit/${id}`} className='list-item'>
    <div>
      <h3 className='list-item__title'>{description}</h3>
      <p className='list-item__sub-title'>{moment(createdAt).format("MMM Do, YYYY")}</p>
    </div>
    <div>
      <h3 className='list-item__data'>{"â‚¹" + numeral(amount).format("0,0.00")}</h3>
    </div>
  </Link>
);

export default ExpenseListItem;
