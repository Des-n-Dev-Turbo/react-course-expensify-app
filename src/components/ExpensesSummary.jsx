import React from "react";
import { connect } from "react-redux";
import { Link } from 'react-router-dom';
import numeral from "numeral";
import getVisibleExpenses from "../selectors/expenses";
import getExpensesTotal from "../selectors/expenses-total";

export const ExpensesSummary = ({ expenseCount, expenseTotal }) => {
  const expenseWord = expenseCount === 1 ? "expense" : "expenses";
  return (
    <div className="page-header">
      <div className="content-container">
        <h1 className='page-header__title'>
          Viewing <span>{expenseCount}</span> {expenseWord} totalling{" "}
          <span>{"â‚¹" + numeral(expenseTotal).format("0,0.00")}</span>
        </h1>
        <div className='page-header__actions'>
            <Link to='/create' className='button'>Add Expense</Link>
        </div>
      </div>
    </div>
  );
};

const mapStateToProps = (state) => {
  const visibleExpense = getVisibleExpenses(state.expenses, state.filters);
  return {
    expenseCount: visibleExpense.length,
    expenseTotal: getExpensesTotal(visibleExpense),
  };
};

export default connect(mapStateToProps)(ExpensesSummary);
