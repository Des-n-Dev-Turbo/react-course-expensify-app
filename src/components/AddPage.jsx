import React from "react";
import { connect } from 'react-redux';
import { addExpense } from "../actions/expenses";
import ExpenseForm from './ExpenseForm';
import { useHistory } from 'react-router-dom';


export const AddPage = (props) => {
  let history = useHistory(); 
  return(
    <div>
      <h2>Add Expense</h2>
      <ExpenseForm 
        onSubmit={(expense) => {
          props.onSubmit(expense);
          history.push('/');
        }}
      />
    </div>
  );
};

const mapDispatchToProps = (dispatch) => ({
  onSubmit: (expense) => dispatch(addExpense(expense))
});

export default connect(undefined, mapDispatchToProps)(AddPage);
