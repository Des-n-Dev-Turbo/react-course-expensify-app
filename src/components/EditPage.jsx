import React from "react";
import {useParams} from 'react-router-dom';
import { useSelector, connect } from 'react-redux';
import ExpenseForm from "./ExpenseForm";
import { editExpense, removeExpense } from "../actions/expenses";
import { useHistory } from "react-router-dom";

export const EditPage = (props) => {
  let history = useHistory();
  let {id} = useParams();
  const expense = useSelector((state) => state.expenses.find((expense) => (expense.id === id)));

  return(
    <div>
      <ExpenseForm
        expense={expense} 
        onSubmit={(expense) => {
          props.onSubmit(id, expense);
          history.push('/');
        }
        }
      />
      <button onClick={
            () => {
                props.onRemove({id});
                history.push('/');
            }
        }>Remove</button>
    </div>
  );
};

// const mapStateToProps = (state) => {
//   let {id} = useParams();
//   return{
//     expense: state.expenses.find((expense) => (expense.id === id))
//   };
// };

const mapDispatchToProps = (dispatch) => ({
  onSubmit : (id, expense) => dispatch(editExpense(id, expense)),
  onRemove: ({id}) => dispatch(removeExpense({id}))
});

export default connect(undefined,mapDispatchToProps)(EditPage);
// export default EditPage;