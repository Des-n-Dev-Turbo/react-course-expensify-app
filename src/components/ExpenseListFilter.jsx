import React from 'react';
import { connect } from 'react-redux';
import { DateRangePicker } from 'react-dates';
import { setTextFilter, sortByAmount, sortByDate, setStartDate, setEndDate } from '../actions/filters';

export class ExpenseListFilter extends React.Component{
    state={
        calendarFocused: null
    };
    onDatesChange = ({ startDate, endDate }) => {
        this.props.setStartDate(startDate);
        this.props.setEndDate(endDate);
    };
    onFocusChange = (calendarFocused) => {
        this.setState(() => ({calendarFocused}));
    };
    onSortChange = (e) => {
        if(e.target.value === 'date'){
            this.props.sortByDate();
        }else if(e.target.value === 'amount'){
            this.props.sortByAmount();
        }
    };
    render(){
        return(
            <div>
                <input type='text' 
                    value={this.props.filters.text} 
                    onChange={(e) => {
                        this.props.setTextFilter(e.target.value);
                    }} 
                />
                <select 
                    value={this.props.filters.sortBy} 
                    onChange={this.onSortChange}
                >
                    <option value='date'>Sort By Date</option>
                    <option value='amount'>Sort By Amount</option>
                </select>
                <DateRangePicker 
                    startDate={this.props.filters.startDate}
                    endDate={this.props.filters.endDate}
                    onDatesChange={this.onDatesChange}
                    focusedInput={this.state.calendarFocused}
                    onFocusChange={this.onFocusChange}
                    numberOfMonths={1}
                    showClearDates={true}
                    isOutsideRange={() => false}
                />
            </div>
        );
    }
};

const mapStateToProps = (state) => ({filters:state.filters});

const mapDispatchToProps = (dispatch) => ({
    sortByDate: () => dispatch(sortByDate()),
    sortByAmount: () => dispatch(sortByAmount()),
    setTextFilter: (value) => dispatch(setTextFilter(value)),
    setStartDate: (startDate) => dispatch(setStartDate(startDate)),
    setEndDate: (endDate) => dispatch(setEndDate(endDate))
});

export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilter);

