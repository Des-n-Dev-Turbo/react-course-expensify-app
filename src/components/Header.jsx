import React from "react";
import {
  NavLink
} from "react-router-dom";
import { connect } from 'react-redux';
import { startLogout } from "../actions/auth";

const Header = ({ startLogout }) => (
  <div>
    <h1>Extensify</h1>
    <NavLink to="/dashboard" exact activeClassName="is-active">
      Home
    </NavLink>
    <NavLink to="/create" activeClassName="is-active">
      Create
    </NavLink>
    <button onClick={startLogout}>Logout</button>
  </div>
);

const mapDispatchToProps = (dispatch) => ({
  startLogout: () => dispatch(startLogout())
});

export default connect(undefined, mapDispatchToProps)(Header);
