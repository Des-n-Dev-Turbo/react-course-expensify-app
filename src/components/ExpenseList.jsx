import React from 'react';
import { connect } from 'react-redux';
import ExpenseListItem from './ExpenseListItem';
import getVisibleExpenses from '../selectors/expenses';




export const ExpenseList = (props) => (
    <div>
        <h2>Expenses</h2>
        {
            props.expenses.length === 0 ? 
            (
                <p>No expenses Available</p>
            ) : (
                props.expenses.map((expense) => (
                <ExpenseListItem key={expense.id} {...expense }/>
                ))
            )
        }
    </div>
);

const mapStateToProps = (state) => {
    return {
        expenses: getVisibleExpenses(state.expenses, state.filters)
    };
};

export default connect(mapStateToProps)(ExpenseList);