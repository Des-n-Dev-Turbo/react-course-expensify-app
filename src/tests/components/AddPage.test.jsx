import React from 'react';
import { shallow } from 'enzyme';
import toJSON from 'enzyme-to-json';
import { AddPage } from '../../components/AddPage';
import expenses from '../fixtures/expenses';

let onSubmit, history, wrapper;

beforeEach(() => {
    onSubmit = jest.fn();
    history = { push : jest.fn() };
    wrapper = shallow(<AddPage onSubmit={onSubmit} history={history} />);
});

test('Should test add page', () => {
    expect(toJSON(wrapper)).toMatchSnapshot();
});

/*
test('Should test onSubmit',() => {
    wrapper.find('ExpenseForm').prop('onSubmit')(expenses[2]);
    expect(history.push).toLastHaveBeenCalledWith('/');
    expect(onSubmit).toLastHaveBeenCalledWith(expenses[2]);
});
*/