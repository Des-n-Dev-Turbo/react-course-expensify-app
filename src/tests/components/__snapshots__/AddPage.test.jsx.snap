// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should test add page 1`] = `
<div>
  <h2>
    Add Expense
  </h2>
  <ExpenseForm
    onSubmit={[Function]}
  />
</div>
`;
